File: fsharp/circular-buffer/CircularBuffer.fs
module CircularBuffer

type RingBuffer(capacity: int) =
    let mutable _items = Array.init capacity (fun i -> 0) 
    member val Capacity = capacity
        with get 
    member val Reader = 0 
        with get, set 
    member val Writer = 0
        with get, set
    member this.Read() = 
        if this.Reader >= this.Writer then 
            raise <| System.Exception()
        else 
            let readIndex = this.Reader % this.Capacity
            this.Reader <- this.Reader + 1
            _items.[readIndex]
    member this.Write(value: int) =
        if (this.Writer - this.Reader = this.Capacity) then 
            raise <| System.Exception()
        else 
            let writeIndex = this.Writer % this.Capacity
            this.Writer <- this.Writer + 1
            _items.[writeIndex] <- value
            ()
    member this.Overwrite(value: int) =
        if (this.Writer - this.Reader < this.Capacity) then 
            this.Write(value)
        else
            let writeIndex = this.Reader % this.Capacity
            this.Writer <- this.Writer + 1
            this.Reader <- this.Reader + 1
            _items.[writeIndex] <- value
            ()

let mkCircularBuffer size = RingBuffer(size)

let clear (buffer: RingBuffer) = RingBuffer(buffer.Capacity)

let write value (buffer: RingBuffer): RingBuffer = 
    buffer.Write(value)
    buffer
        
let forceWrite value (buffer: RingBuffer) = 
    buffer.Overwrite(value)
    buffer

let read (buffer: RingBuffer): (int * RingBuffer) = 
    let value = buffer.Read()
    value, buffer

File: csharp/circular-buffer/CircularBuffer.cs
using System;
using System.Collections.Generic;

public class CircularBuffer<T>
{
    private int _capacity;
    private T[] _buffer;
    private int _reader = 0;
    private int _writer = 0;
    public CircularBuffer(int capacity)
    {
        _capacity = capacity;
        _buffer = new T[_capacity]; 
    }

    public T Read()
    {
        if (_reader >= _writer) throw new InvalidOperationException();
        int indexToRead = _reader % _capacity;
        _reader++;
        return _buffer[indexToRead];
    }

    public void Write(T value)
    {
        if (_writer - _reader == _capacity) throw new InvalidOperationException();
        int indexToWrite = _writer % _capacity;
        _writer++;
        _buffer[indexToWrite] = value;
        return;
    }

    public void Overwrite(T value)
    {
        if (_writer - _reader < _capacity)
        {
            Write(value);
            return;
        }
        int indexToWrite = _reader % _capacity;
        _reader++;
        _writer++;
        _buffer[indexToWrite] = value;
        return;
    }

    public void Clear()
    {
        _reader = 0;
        _writer = 0;
        _buffer.Initialize();
        return;
    }
}